docs/proposed/buildtime-and-runtime-detection-for-hardware-acceleration.md

```diff
@@ -0,0 +1,245 @@
+# Build-time and runtime detection for hardware acceleration with advance CPU features
+
+This document describes how to support the algorithms that can do hardware acceleration with advance CPU features.
+
+## Introduction
+
+As description in [Build on Arm64 HOST Fail](https://github.com/Mbed-TLS/mbedtls/issues/5758), we got two issues that can be fixed with [#6932](https://github.com/Mbed-TLS/mbedtls/pull/6932)
+- `check_config.h` reports fail when HOST-CC is Arm64
+- With `-march=armv8.2-a+sha3` cpu flags
+Until [#6895](https://github.com/Mbed-TLS/mbedtls/pull/6895), we forbiden enable `MBEDTLS_SHA512_USE_A64_IF_PRESENT` and `MBEDTLS_AESCE_C` to workaround it.
+
+For time being, we have 4 hardware accelerations in `development` and 1 in [#6918](https://github.com/Mbed-TLS/mbedtls/pull/6918)
+- SHA256 with Arm64 crypto extension
+- SHA512 with Arm64 sha3 extension
+- AES/GCM with x86_64 aes extension
+- AES/GCM with x86 padlock extension
+- AES/GCM with Arm64 crypto extension
+There are 5 functions for `runtime detection`. We can remove two functions. For Arm64, the detection function in SHA512 is the best one.
+And SHA{512,256} implement Arm64 hardware acceleration only. AES/GCM does not implmented it
+
+As my understand, we should meet below requirements.
+- With `runtime detection`.
+  - `runtime detection` for architecture MUST be enabled.
+  - Enable/disable hardware accelerations with configure options.
+  - Enable/disable `no hardware accelerations` with configure options.
+- Without `runtime detection`( Build-time detection only).
+  - `runtime detection` for architecture MUST be disabled.
+  - If `no hardware accelerations` is enabled, all hardwares accelerations MUST be disabled.
+  - If `no hardware accelerations` is disabled, only one hardwares acceleration can be enabled.
+  - If possible, enable/disable hardware accelerations base on `CFLAGS`.
+
+NOTE: The algorithm can be implemented with different instruction sets on same architecture. For example, AES can be implemented with crypto extention or SVEAES extension in Arm64.
+
+
+## Suggestion
+
+### check config
+
+`check_config.h` MUST not include any archeticture special checks. This gives up the advantage of the current code. But that makes it architecture independent.
+
+For `build-time detection only`, we can enable the the checks in hardware acceleration module.
+
+There are a potential issue for `build-time detection only`, if library is built with some CPU modifer flags, how application known the parameter? Is it important ?
+
+### Configure options
+
+- `MBEDTLS_RUNTIME_DETECTION`: Enable runtime detection feature.
```

Comment:
> Should this really be a configuration option? 

Yes. But I think it should be renamed to `MBEDTLS_RUNTIME_C` now.

```c
static int aes_internal_setkey_enc(unsigned char *rk,
                                   const unsigned char *key,
                                   size_t bits)
{

#if defined(MBEDTLS_ARCH_IS_ARM64)
#if defined(MBEDTLS_AESCE_C) 
    if (mbedtls_aesce_has_support()) {
        return mbedtls_aesce_setkey_enc(rk, key, bits);
    }
#endif
#endif /* MBEDTLS_ARCH_IS_ARM64 */

#if !defined(MBEDTLS_AES_DISABLE_NO_ACCEL)
    return aes_setkey_enc_no_accel((uint32_t *) rk, key, bits);
#else
    return 0;
#endif

}
```

Reply:
What I don't understand is why it should be a configuration option. Each configuration option increases the cost of testing. Why can't we always have `mbedtls_cpu_has_feature`, and it's a `static inline` constant function (so no cost in code size) if nothing needs the information?

Reply:
> it's a `static inline` constant function (so no cost in code size) if nothing needs the information?

It has cost in code size. The differences are the code size on different architecture.  With `MBEDTLS_RUNTIME_C`, we can just enable it with `#define mbedtls_cpu_has_feature(...) 1`,  the code size from `runtime detection` will be reduced. 

For x86/x64, the cost in code size is very small. from the [code size comparision](https://godbolt.org/z/xG4v9sYc9) , 14 extra instructions was generated. 

For Arm64/32, the issue is very complex. `sha512.c` shows the solutions. It is not small cost operation.
- With elf hwcap ( linux), the cost is small than x86/x64. 
- Apple/Windows, I guess it will be trapped into kernel mode. It is heavy cost for runtime function. It needs a profile function for it. It is a big cost in the code size.
- For others, we can profile the features 
  - profile with SIG_ILL signal
  - profile with reading sys registers if it is allowed.

Reply:
I don't think this config option is needed either. In the corresponding WIP PR there are macros to detect whether the runtime config option is needed and error out if it is not present when needed or present when not needed:
https://github.com/Mbed-TLS/mbedtls/pull/7078/commits/32ac20ac047a91d8442b169036d68622bb3cf7b9#diff-4fc3811c7c53e1746e1056ac38a6462ad2160077a59898bbd2bc31f5938ac185R1101-R1116

How about instead of erroring out just turning `MBEDTLS_RUNTIME_C` on and off as needed and not having it as a config option in `config.h` at all?

Reply:
>Why can't we always have mbedtls_cpu_has_feature, and it's a static inline constant function (so no cost in code size) if nothing needs the information?

@gilles-peskine-arm , do you mean implement like bellow
```c
static inline int mbedtls_cpu_has_features(uint32_t flags)
{
#if defined(MBEDTLS_RUNTIME_C)
        return mbedtls_cpu_has_features_arm64(flags);
#else
       return (MBEDTLS_TARGET_CPU_FEATURES&flags)==flags;
#endif
}
```
I think I misunderstand your meaning.

Reply:
I just add a [table](https://github.com/Mbed-TLS/mbedtls/pull/7004/commits/c5324f3ed29906f165db15c726f2e6d908756558) for macros.

Reply:
I have update #7078 to remove many configuration options.   I guess it is suggestion from @gilles-peskine-arm  now.

Reply:
No, I mean something like:
```
static inline int mbedtls_cpu_has_features(uint32_t flags) {
#if defined(MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT) || defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) || defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT)
    return mbedtls_cpu_has_features_arm64(flags);
#else
    return 0;
#endif
}
```
I don't see a need for a separate compile-time option.

Reply:
I agree that we don't need `MBEDTLS_RUNTIME_C` in as a user facing high level compile time option in `mbedtls_config.h`.

We can have `MBEDTLS_RUNTIME_C` macro defined somewhere else (eg `build_info.h` or `runtime.h`) as a helper macro to improve readability and for convenience:

```C
#if defined(MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT)    ||  \
    defined(MBEDTLS_SHA256_USE_A64_CRYPTO_IF_PRESENT) ||  \
    defined(MBEDTLS_SHA512_USE_A64_CRYPTO_IF_PRESENT)
#define MBEDTLS_RUNTIME_C
#endif
}
```

Reply:
Understand. 

And your example code is wrong.  It should be
```c
#if defined(MBEDTLS_AES_CE_C)    ||  \
    defined(MBEDTL_AES_SVE_C) ||  \
    defined(MBEDTLS_AES_PURE_SOFTWARE_C)
#define MBEDTLS_RUNTIME_C
#endif
```

If any algorithm enables two accelerator , we should enable runtime detection.

Reply:
Indeed, my example code was still using the old 3 state macros, in the new proposal there wouldn't be any `MBEDTLS_AES_USE_A64_CRYPTO_IF_PRESENT` macros at all.

Reply:
I mean, if an algorithm has more than 1 accelerators , `MBEDTLS_RUNTIME_C` should be enabed. Not base on different algorithms. And I update my example
```c
#if MBEDTLS_AES_USE_A64_CRYPO + MBEDTLS_AES_USE_A64_SVE + MBEDTLS_AES_USE_PURE_SOFTWARE_C > 1
#define MBEDTLS_RUNTIME_C
#elif MBEDTLS_AES_USE_A64_CRYPO + MBEDTLS_AES_USE_A64_SVE + MBEDTLS_AES_USE_PURE_SOFTWARE_C == 1
/* Runtime detection was disbaled */
#else
#error "MBEDTLS_AES_C defined, But no accelerator provided"
#endif

#if MBEDTLS_SHA256_USE_A64_CRYPO + MBEDTLS_SHA256_USE_A64_SVE + MBEDTLS_SHA256_USE_PURE_SOFTWARE_C > 1
#define MBEDTLS_RUNTIME_C
#elif MBEDTLS_SHA256_USE_A64_CRYPO + MBEDTLS_SHA256_USE_A64_SVE + MBEDTLS_SHA256_USE_PURE_SOFTWARE_C == 1
/* Runtime detection was disbaled */
#else
#error "MBEDTLS_SHA256_C defined, But no accelerator provided"
#endif
```

Reply:
Done in #7078. Please double check if it is expected.